README.rst
.. This README is meant for consumption by humans and pypi. Pypi can render rst files so please do not use Sphinx features.
   If you want to learn more about writing documentation, please check out: http://docs.plone.org/about/documentation_styleguide.html
   This text does not appear on pypi or github. It is a comment.

.. image::
   https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336
       :target: https://pycqa.github.io/isort/

====================
visaplan.plone.tools
====================

General tools modules for Plone_.

We don't claim ultimate Plone wisdom (yet);
this package is one of the parts a big monolithic classic Zope product
was split into.

It is part of the footing of the "Unitracc family" of Plone sites
which are maintained by `visaplan GmbH`_, Bochum, Germany.

Some modules of this package might still contain some resources
(e.g. type names)
which are specific to our "Unitracc family" of sites;
this is likely to change in future releases.


Features
--------

This add-on can be seen in action at the following sites:

- https://www.unitracc.de
- https://www.unitracc.com

Modules in version 1.1.4+:

- ``attools`` module

  several tools for Archetypes-based content

- ``brains`` module

  currently one ``make_collector`` function, e.g. for address fields

- ``cfg`` module

  Read "product" configuration, and detect development mode

- ``context`` module

  Several tools for processing the request.
  Some need some modernization ...

- ``forms`` module

  Several tools for forms

- ``functions`` module

  Some functions, e.g. ``is_uid_shaped``

- ``indexes`` module (new in v1.1.4) 

  - Function ``getSortableTitle`` for title conversion.

    This converts umlauts etc. to sort them
    as equal to their corresponding base vocals,
    according to German lexical usage.

- ``log`` module

  Automatically named loggers

- ``mock`` module

  - a few small classes for use in doctests

  - the same module as ``visaplan.tools.mock``

- ``mock_cfg`` module

  A rudimentary mock module for ``cfg``

- ``search`` module

  A few functions to support creation of ZODB catalog search queries
  (quite proprietary, I'm afraid; might go away in future versions)

- ``setup`` module (since v1.1)

  Functions for use in migration scripts

- ``zcmlgen`` module (since v1.1.1)

  - Generates ``configure.zcml`` files, if

    - changes are detected, and

    - development mode is active, and

    - the source is in an development package.

- ``decorators`` module (since v1.1.6)

  - ``@returns_json``:

    Wraps the function call and returns the JSON_-encoded result,
    including HTTP headers.

    Uses simplejson_ if available.

Documentation
-------------

Sorry, we don't have real user documentation yet.

Most functions are documented by doctests, anyway;
it helps to understand some German.


Installation
------------

Since ``visaplan.plone.tools`` is a package for Plone instances,
it is not normally installed using ``pip``;
instead, install it by adding it to your buildout::

    [buildout]
    ...

    eggs =
        visaplan.plone.tools


and then running ``bin/buildout``


Contribute
----------

- Issue Tracker: https://github.com/visaplan/plone.tools/issues
- Source Code: https://github.com/visaplan/plone.tools


Support
-------

If you are having issues, please let us know;
please use the `issue tracker`_ mentioned above.


License
-------

The project is licensed under the GPLv2.

.. _`issue tracker`: https://github.com/visaplan/plone.tools/issues
.. _JSON: https://json.org/
.. _Plone: https://plone.org/
.. _simplejson: https://pypi.org/project/simplejson
.. _`visaplan GmbH`: http://visaplan.com

.. vim: tw=79 cc=+1 sw=4 sts=4 si et


CHANGES.rst
.. vim: sw=2 sts=2 tw=79 cc=+1

Changelog
=========


1.4.15 (unreleased)
-------------------

- Py3K support fixes for

  - .functions.is_uid_shaped
  - .log.getLogSupport

[tobiasherp]


1.4.14 (2022-04-22)
-------------------

Bugfixes:

- Missing imports fixed for

  - .groups.is_member_of__factory
  - .setup.get_default_idxs

[tobiasherp]


1.4.13 (2022-03-04)
-------------------

Bugfixes:

- Fixed a bug in .groups.groupinfo_factory(pretty);
  for this to work, we need visaplan.plone.groups.

Improvements:

- The label of the default logger of the ``@@step`` decorator
  now includes the name of the worker, e.g. ``instance`` or ``client``

New features: 

- New `.env` module, providing the `worker_name()`

[tobiasherp]


1.4.12 (2022-02-15)
-------------------

Bugfixes:

- Fixed a bug in .groups.groupinfo_factory(pretty);
  for this to work, we need visaplan.plone.groups.

[tobiasherp]


1.4.11 (2022-02-05)
-------------------

Bugfixes:

- Fixed a regression for .groups.groupinfo_factory

[tobiasherp]


1.4.10 (2022-02-03)
-------------------

Improvements:

- Converted the .groups module in a subpackage

New features: 

- New option `missing=False` for

  - .groups.groupinfo_factory

  If `True`, the resulting function creates an `existing` key,
  and for missing groups, the `group_title` is `None`;
  otherwise, an empty dict is returned in such cases.

[tobiasherp]


1.4.9 (2021-12-13)
------------------

Improvements:

- If collective.metadataversion_ is installed, use the configured default set of
  indexes (if no `idxs` explicitly given but metadata recreation is requested)

[tobiasherp]


1.4.8 (2021-11-29)
------------------

New features: 

- New ``.setup`` function `load_and_cook`, for use in GenericSetup upgrade steps

[tobiasherp]


1.4.7 (2021-11-17)
------------------

Bugfixes:

- When detecting the currently active language (``.context.getActiveLanguage[_unchecked]``),
  we now check for the ``I18N_LANGUAGE`` cookie first.

[tobiasherp]


1.4.6 (2021-10-28)
------------------

Bugfixes:

- Fixed an error for the new functions in the ``.attools`` and ``.dxtools`` modules
  when giving the schema as positional argument
  (which to use isn't implemented yet for ``.dxtools`` anyway).

[tobiasherp]


1.4.5 (2021-10-27)
------------------

(If using the new functions in the .axtools  module, please use release
1.4.6+ instead).

Breaking changes:

- Removed the .metadata module which had been added in release 1.4.2;
  we have collective.metadataversion_ now to replace it,
  storing the `metadata_version`_ value persistently in the registry.

New features: 

- New .attools functions:

  - `get_first_text_as_html`
  - `get_all_texts`
  - `generate_all_texts`

- New .dxtools module, containing the functions:

  - `get_first_text_as_html`
  - `get_all_texts`
  - `generate_all_texts`

Profile changes:

- Removed the ``default`` profile;
  there is currently no point in installing this package as a Plone plugin.
  Just add it to your required eggs, and import from the modules.

  We keep the the ``uninstall`` profile *for now;*
  it will be removed in an near-future version.

  So, *don't "install"* this package (Quick-Installer, Plone add-ons);
  just use it in Python_ code!

  We keep the ``configure.zcml`` file and the autoinclude entry point, though;
  we can image to use e.g. the Plone registry for some settings.

[tobiasherp]


1.4.4 (2021-08-31)
------------------

Bugfixes:

- .setup.make_object_getter() didn't update the Language index when the language was changed

Improvements:

- .setup.make_object_getter() now additionally updates the following indexes
  when the .title attribute is changed:

  - sortable_title
  - SearchableText
  - getEffectiveIndex

[tobiasherp]


1.4.3 (2021-08-27)
------------------

New features: 

- ``.search`` module:

  - New conversion utilities ...

    - `make_querystring_mangle`, a factory, to create
    - `mangleQueryString`, using
    - `mangle_umlauts`

Miscellaneous:

- `normalizeQueryString` (as well as the new `mangleQueryString`)
   now auto-inserts asterisks at word ends *only*

[tobiasherp]


1.4.2 (2021-08-25)
------------------

New utilities:

- ``.context.getActiveLanguage_unchecked``: like `getActiveLanguage`,
  but not checking against the supported languages, and thus faster
  (e.g. for cache key functions)

Temporary changes:

- New ``.metadata`` module to support conditional metadata updates.

  **Note:** this is removed in release 1.4.5.
  If you need the `metadata_version`_ metadata column, please use collective.metadataversion_ instead.

[tobiasherp]


1.4.1 (2021-06-30)
------------------

Bugfixes:

- Bugfix for ``.setup.handle_subportal``

Hints:

- visaplan.plone.search contains a `make_input_text` function
  in it's ``.utils`` module
  which re-implements `make_input` without using BeautifulSoup.
  This will probably replace our present `make_input` function and then
  disregard any keyword options which are currently passed on to the
  BeautifulSoup_ constructor, i.e. the `factory` option.

  Perhaps we'll create a new visaplan.zope.tools package and move this
  functionality there, though.

[tobiasherp]


1.4.0 (2021-03-26)
------------------

Breaking changes:

- `.setup.switch_menu_item` ...

  - won't change anything if the `on` argument is `None`
  - won't accept (other) non-boolean arguments for `on`,
    unless `strict=False` is specified (new keyword-only argument)

Improvements:

- Working doctests for ``search`` module
- `zcmlgen` constructors support `skip` option

Requirements:

- visaplan.tools_ v1.3.1+

[tobiasherp]


1.3.0 (2020-12-16)
------------------

New Features:

- New module ``groups``

- New `.context` functions

  - `getMessenger` (factory):

    creates a `message` function which doesn't require
    (nor accept) a `context` argument

  - `getPath`
  - `get_parent`
  - `parents`
  - `parent_brains`
  - `make_brainGetter`
  - `make_pathByUIDGetter`
  - `make_translator`

src/visaplan/plone/tools/_at2.py
# -*- coding: utf-8 -*- vim: ts=8 sts=4 sw=4 si et tw=79
"""
visaplan.plone.tools._at2 -- imported via .attools
"""

# Python compatibility:
from __future__ import absolute_import, print_function

from six import string_types as six_string_types

# Standard library:
from collections import defaultdict

# visaplan:
from visaplan.tools.lands0 import makeSet

# "Log-Level" für generierte Funktion:
BASE, CHANGE, NOP, SKIP, DEBUG = (1, 2, 3, 4, 99)

CORE_FIELDS = set([
        'id', 'title',
        ])
BOOLEAN_FIELDS = set([
        'excludeFromNav',
        'allowDiscussion',
        'excludeFromSearch',
        ])
DATE_FIELDS = set([
        'creation_date', 'modification_date',
        'effectiveDate', 'expirationDate',
        ])
UNITRACC_NUMERIC_FIELDS = set([
        'price_shop',
        'shopduration',
        ])
UNITRACC_CHOICES_FIELDS = set([
        'code',
        'unitraccGroups',
        'subPortals',
        ])
PLONE_OTHER_FIELDS = set([  # sonstige (Plone):
        'creators',
        'language',
        ])
UNITRACC_OTHER_FIELDS = set([
        # XML:
        'xml',
        # evtl. später doch berücksichtigen :
        'certificate',
        'contributors',
        'difficulty',
        'duration',
        'html_left',
        'html_right',
        'location',
        'price',
        'relatedItems',
        'subject',
        'target_group',
        ])
# Nicht nach HTML konvertieren:
DEFAULT_BLACKLIST = set()
DEFAULT_BLACKLIST.update(CORE_FIELDS)
DEFAULT_BLACKLIST.update(BOOLEAN_FIELDS)
DEFAULT_BLACKLIST.update(DATE_FIELDS)
DEFAULT_BLACKLIST.update(UNITRACC_NUMERIC_FIELDS)
DEFAULT_BLACKLIST.update(UNITRACC_CHOICES_FIELDS)
DEFAULT_BLACKLIST.update(PLONE_OTHER_FIELDS)
DEFAULT_BLACKLIST.update(UNITRACC_OTHER_FIELDS)


def make_minilogger(verbose, logger, prefix=None):
    """
    Erzeuge den Minilogger, üblicherweise tell(LEVEL, text[, mapping ...])
    """
    def make_mapping(*args):
        if not args:
            return {}
        elif (not args[1:]) and isinstance(args[0], dict):
            return args[0]
        else:
            return args

    if verbose:
        if logger is None:
            if prefix:
                MASK = prefix.replace('%', '%%') + ': %s'
                def _tell(s, *args):
                    print(MASK % (s % make_mapping(*args)))
            else:
                def _tell(s, *args):
                    print(s % make_mapping(*args))
        else:
            _tell = logger.info
        def tell(level, s, *args):
            if verbose >= level:
                _tell(s, *args)
                return True
    else:
        def tell(level, s, *args):
            pass
    return tell


def make_skip_function(tell, **kwargs):
    """
    Gib eine skip-Funktion zurück, die Attributnamen mit der black- oder
    whitelist vergleicht
    """
    if ('sets' in kwargs
        and 'whitelist' not in kwargs
        and 'blacklist' not in kwargs
        ):
        sets = kwargs.pop('sets')
        if sets:
            def skip(fieldname):
                return fieldname not in sets
            return skip

    whitelist = kwargs.pop('whitelist', None)
    blacklist = kwargs.pop('blacklist', None)
    if whitelist is not None:
        assert blacklist is None
        if isinstance(whitelist, six_string_types):
            whitelist = [whitelist]
        whitelist = set(whitelist)
        def skip(fieldname):
            if fieldname not in whitelist:
                tell(SKIP, '%(fieldname)r not in whitelist --> SKIP', locals())
                return True
            return False
    else:
        if blacklist is None:
            blacklist = []
        elif isinstance(blacklist, six_string_types):
            blacklist = [blacklist]
        blacklist = set(blacklist)
        blacklist.update(DEFAULT_BLACKLIST)
        def skip(fieldname):
            if fieldname in blacklist:
                tell(SKIP, '%(fieldname)r in blacklist --> SKIP', locals())
                return True
            return False
    return skip


def make_mimetype_fixer(**kwargs):
    """
    Erzeuge eine Funktion, die das übergebene Archetypes-Objekt korrigiert.

    Alle Optionen müssen als Schlüsselwortargument übergeben werden:

    from_mimetype  -- Vorgabe: keine Filterung
    to_mimetype -- Vorgabe: 'text/html'
    default_output_type -- Vorgabe: 'text/x-html-safe'

    counter -- üblicherweise ein defaultdict(int).  Wenn nicht übergeben, wird
               eines erzeugt, dessen Werte dann nicht zugänglich sind.
    verbose -- ausführliche Ausgaben/Logging?
    logger -- der zu verwendende Logger
    """
    # --------------------------- [ Argumente für die Ausführung ... [
    verbose = kwargs.pop('verbose', None)
    logger = kwargs.pop('logger', None)
    if verbose is None:
        verbose = logger is not None
    tell = make_minilogger(verbose, logger, 'fix_types')

    if 'counter' in kwargs:
        counter = kwargs.pop('counter')
    else:
        counter = defaultdict(int)

    # zu bearbeitende Attribute:
    skip = make_skip_function(tell, **kwargs)
    # --------------------------- ] ... Argumente für die Ausführung ]

    # --------------------------- [ Argumente für die Konversion ... [
    from_mimetypes = []
    from_mimetype = kwargs.pop('from_mimetype', None)
    if from_mimetype is None:
        from_mimetypes = None
    elif isinstance(from_mimetype, six_string_types):
        from_mimetypes.append(from_mimetype)
    else:
        from_mimetypes.extend(list(from_mimetype))

    to_mimetype = kwargs.pop('to_mimetype', 'text/html')
    assert to_mimetype, ('to_mimetype wird benoetigt! (%(to_mimetype)r)'
                         ) % locals()

    if 'default_output_type' in kwargs:
        default_output_type = kwargs.pop('default_output_type')
    elif to_mimetype == 'text/html':
        default_output_type = 'text/x-html-safe'
    else:
        default_output_type = None
    # --------------------------- ] ... Argumente für die Konversion ]

    assert not kwargs, 'Nicht erkannte Argumente: %s' % (kwargs,)

    def fix_types(instance=None, brain=None):
        # Docstring-Generierung: siehe unten
        instance, iinfo = instance_and_label(instance, brain, verbose)
        schema = instance.Schema()
        changed = False
        if verbose:
            tell(BASE, 'fixing types for %(iinfo)s ...', locals())
        for field in schema.values():
            name = field.__name__
            if skip(name):
                continue
            tell(DEBUG, 'field: name=%(name)r,'
                        ' field=%(field)r', locals())
            # nur schreibbare Felder:
            try:
                mutator = field.getMutator(instance)
                if mutator is None:
                    counter[('skipped', name, 'readonly')] += 1
                    tell(SKIP, 'field %(name)r is readonly', locals())
                    continue
            except AttributeError as e:
                counter[('skipped', name, 'AttributeError', 'getMutator')] += 1
                tell(BASE, 'AttributeError: getMutator')
                continue
            try:
                base_unit = field.getBaseUnit(instance)
            except AttributeError as e:
                counter[('skipped', name, 'AttributeError', 'getBaseUnit')] += 1
                tell(BASE, 'AttributeError: getBaseUnit')
                continue
            tell(DEBUG, 'OK: Feld %(name)r hat Mutator und BaseUnit', locals())
            mimetype = base_unit.mimetype
            if (from_mimetypes is not None
                and mimetype not in from_mimetypes
                ):
                tell(SKIP, 'skipping %(mimetype)r field %(name)r', locals())
                counter[('skipped', name, 'type', mimetype)] += 1
                continue
            args = []
            kwargs = {}
            if not base_unit:
                args.append(field.getDefault(instance))
                tell(CHANGE, 'creating field %(name)r', locals())
                kwargs['_initializing_'] = True
            if mimetype != to_mimetype or not base_unit:
                kwargs['mimetype'] = to_mimetype
            if (default_output_type
                    and field.default_output_type != default_output_type):
                kwargs['default_output_type'] = default_output_type
            if args or kwargs:
                kwargs.update({'field': name,
                               })
                if not args:
                    tell(DEBUG, 'Getting existing value for %(name)r ...', locals())
                    oldval = base_unit.getRaw()
                    args.append(oldval)
                    tell(CHANGE, 'using raw value %(oldval)r for field %(name)r', locals())
                tell(CHANGE, 'rectifying field %(name)r', locals())
                counter[('change', name)] += 1
                mapply(mutator, *args, **kwargs)
                changed = True
            else:
                tell(NOP, 'checked field %(name)r', locals())

        if changed:
            counter['changed'] += 1
            if verbose:
                tell(BASE, 'fixing types for %(iinfo)s: DONE', locals())
        else:
            counter['unchanged'] += 1
            if verbose:
                tell(BASE, 'fixing types for %(iinfo)s: NOTHING TO DO', locals())
        return instance

    fix_types.__doc__ = """
        Korrigiere MIME-Typen für das übergebene Objekt

        from_mimetypes =      %(from_mimetypes)s
        to_mimetype =         %(to_mimetype)r
        default_output_type = %(default_output_type)r
        """ % locals()
    return fix_types


def make_fields_inspector(**kwargs):
    """
    Erzeuge eine Funktion, die Informationen über bestimmte Felder ausgibt
    """
    # --------------------------- [ Argumente für die Ausführung ... [
    verbose = kwargs.pop('verbose', True)
    logger = kwargs.pop('logger', None)
    if verbose is None:
        verbose = logger is not None
    tell = make_minilogger(verbose, logger, 'inspect')

    if 'counter' in kwargs:
        counter = kwargs.pop('counter')
    else:
        counter = defaultdict(int)

    # zu inspizierende Attribute:
    fieldnames_good = makeSet(kwargs.pop('fieldnames_good', None))
    fieldnames_broken = makeSet(kwargs.pop('fieldnames_broken', None))
    skip = make_skip_function(tell,
                              sets=fieldnames_good+fieldnames_broken,
                              **kwargs)
    # --------------------------- ] ... Argumente für die Ausführung ]
    def this_cat(name):
        if name in fieldnames_good:
            return 2
        if name in fieldnames_broken:
            return 1
        return 0

    cat_label = ['Attribute ohne Einordnung:',
                 'Negativ-Beispiele:',
                 'Positiv-Beispiele:',
                 ]
    field_attributes = ['metatype',
                        'mimetype',
                        'default_output_type',
                        ]
    baseunit_attributes = ['mimetype',
                           ]

    def inspect_field(field, instance):
        """
        Funktion in Entwicklung ...
        """
        tell(BASE, '* Feld %r (%s):', field.__name__, field)
        has_attrs = set()
        lacks_attrs = set()
        field_value = {}
        for attr_name in field_attributes:
            try:
                attr = getattr(field, attr_name)
                has_attrs.add(attr_name)
                field_value[attr_name] = attr
                tell(BASE, '  %(attr_name)s: %(attr)r', locals())
            except AttributeError:
                lacks_attrs.add(attr_name)
        try:
            base_unit = field.getBaseUnit(instance)
            tell(BASE, '* %(base_unit)r:', locals())
        except AttributeError as e:
            tell(BASE, '  (keine BaseUnit)')
        else:
            for attr_name in baseunit_attributes:
                try:
                    attr = getattr(base_unit, attr_name)
                    if attr_name in has_attrs:
                        if attr == field_value[attr_name]:
                            suffix = ' (gleich)'
                        else:
                            suffix = ' (UNGLEICH)'
                    else:
                        suffix = ''
                    tell(BASE, '  - %(attr_name)s: %(attr)r%(suffix)s', locals())
                except AttributeError:
                    pass
            raw = base_unit.getRaw()
            tell(BASE, '  = %r', raw[:30])

    def inspect(instance=None, brain=None):
        instance, iinfo = instance_and_label(instance, brain, verbose)
        schema = instance.Schema()
        prev_cat = None
        for field in schema.values():
            name = field.__name__
            if skip(name):
                continue
            if prev_cat is None:
                tell(iinfo)
            curr_cat = this_cat(name)
            if curr_cat != prev_cat:
                tell(cat_label[curr_cat])
                prev_cat = curr_cat

            inspect_field(field, instance)
